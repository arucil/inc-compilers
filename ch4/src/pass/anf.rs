use ast::{Exp, IdxVar, Program};
use support::Range;

pub fn anf(prog: Program<IdxVar>) -> Program<IdxVar> {
  let mut counter = 0;
  Program {
    body: prog
      .body
      .into_iter()
      .map(|(range, exp)| (range, anf_exp(range, exp, &mut counter)))
      .collect(),
  }
}

fn anf_exp(range: Range, exp: Exp<IdxVar>, counter: &mut usize) -> Exp<IdxVar> {
  match exp {
    Exp::Let { var, init, body } => Exp::Let {
      var,
      init: box (init.0, anf_exp(init.0, init.1, counter)),
      body: box (body.0, anf_exp(body.0, body.1, counter)),
    },
    Exp::Prim { op, args } => {
      let mut tmps = vec![];
      let args = args
        .into_iter()
        .map(|(range, arg)| (range, atom_exp(range, arg, &mut tmps, counter)))
        .collect();
      tmps
        .into_iter()
        .rfold(
          (range, Exp::Prim { op, args }),
          |(body_range, body), (range, var, init)| {
            (
              body_range,
              Exp::Let {
                var: (range, var),
                init: box (range, init),
                body: box (body_range, body),
              },
            )
          },
        )
        .1
    }
    Exp::If { cond, conseq, alt } => Exp::If {
      cond: box (cond.0, anf_exp(cond.0, cond.1, counter)),
      conseq: box (conseq.0, anf_exp(conseq.0, conseq.1, counter)),
      alt: box (alt.0, anf_exp(alt.0, alt.1, counter)),
    },
    Exp::Set { var, exp } => Exp::Set {
      var,
      exp: box (exp.0, anf_exp(exp.0, exp.1, counter)),
    },
    Exp::Begin { seq, last } => Exp::Begin {
      seq: seq
        .into_iter()
        .map(|exp| (exp.0, anf_exp(exp.0, exp.1, counter)))
        .collect(),
      last: box (last.0, anf_exp(last.0, last.1, counter)),
    },
    Exp::While { cond, body } => Exp::While {
      cond: box (cond.0, anf_exp(cond.0, cond.1, counter)),
      body: box (body.0, anf_exp(body.0, body.1, counter)),
    },
    Exp::Print { val, ty } => Exp::Print {
      val: box (val.0, anf_exp(val.0, val.1, counter)),
      ty,
    },
    exp => exp,
  }
}

fn atom_exp(
  range: Range,
  exp: Exp<IdxVar>,
  tmps: &mut Vec<(Range, IdxVar, Exp<IdxVar>)>,
  counter: &mut usize,
) -> Exp<IdxVar> {
  match exp {
    Exp::Prim { op, args } => {
      let args = args
        .into_iter()
        .map(|(range, arg)| (range, atom_exp(range, arg, tmps, counter)))
        .collect();
      let exp = Exp::Prim { op, args };
      assign_var(range, exp, tmps, counter)
    }
    Exp::Let { .. }
    | Exp::If { .. }
    | Exp::Set { .. }
    | Exp::Begin { .. }
    | Exp::While { .. } => {
      let exp = anf_exp(range, exp, counter);
      assign_var(range, exp, tmps, counter)
    }
    exp => exp,
  }
}

fn assign_var(
  range: Range,
  exp: Exp<IdxVar>,
  tmps: &mut Vec<(Range, IdxVar, Exp<IdxVar>)>,
  counter: &mut usize,
) -> Exp<IdxVar> {
  let tmp = IdxVar {
    name: "tmp".to_owned(),
    index: *counter,
  };
  *counter += 1;
  tmps.push((range, tmp.clone(), exp));
  Exp::Var(tmp)
}

#[cfg(test)]
mod tests {
  use super::*;
  use ast::*;
  use insta::assert_snapshot;

  #[test]
  fn nested_prims() {
    let prog =
      parse(r#"(let ([x (read)] [y (+ 2 3)]) (+ (- (read)) (+ y (- 2))))"#)
        .unwrap();
    let prog = super::super::uniquify::uniquify(prog);
    let result = anf(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn init_with_var() {
    let prog = parse(r#"(let ([a 42]) (let ([b a]) b))"#).unwrap();
    let prog = super::super::uniquify::uniquify(prog);
    let result = anf(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn let_in_init() {
    let prog = parse(
      r#"(let ([a (+ (let ([x (read)] [y (- x)]) (+ x (- y))) 7)]) (- a))"#,
    )
    .unwrap();
    let prog = super::super::uniquify::uniquify(prog);
    let result = anf(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn if_form() {
    let prog = parse(
      r#"(+ (- 3 2) (if (not (eq? 2 3)) (- (read)) (> (read) (not #t))))"#,
    )
    .unwrap();
    let prog = super::super::uniquify::uniquify(prog);
    let result = anf(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn begin() {
    let prog = parse(
      r#"
(let ([x0 10])
  (let ([y1 0])
    (+ (+ (begin (set! y1 (read)) x0)
       (begin (set! x0 (read)) y1))
    x0)))"#,
    )
    .unwrap();
    let prog = super::super::uniquify::uniquify(prog);
    let result = anf(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
