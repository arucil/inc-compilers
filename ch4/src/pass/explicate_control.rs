use asm::Label;
use ast::{Exp, ExpKind, IdxVar, Program, Type};
use control::{CAtom, CCmpOp, CExp, CPrim, CProgram, CStmt, CTail};
use indexmap::{IndexMap, IndexSet};
use std::collections::VecDeque;
use std::fmt::{self, Debug, Formatter};

pub struct CInfo {
  pub locals: IndexSet<IdxVar>,
}

impl Debug for CInfo {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    writeln!(f, "locals: {:?}\n", self.locals)
  }
}

struct State {
  blocks: Vec<(Label, CTail)>,
  label_index: u32,
}

impl State {
  fn new_label(&mut self) -> Label {
    let label = Label::Tmp(self.label_index);
    self.label_index += 1;
    label
  }

  fn add_label_block(&mut self, label: Label, block: CTail) {
    self.blocks.push((label, block));
  }

  fn add_block(&mut self, block: CTail) -> Label {
    let label = self.new_label();
    self.add_label_block(label, block);
    label
  }

  fn tail_to_label(&mut self, tail: CTail) -> Label {
    match tail {
      CTail::Goto(label) => label,
      _ => self.add_block(tail),
    }
  }

  fn tail_to_goto(&mut self, tail: CTail) -> CTail {
    match tail {
      CTail::Goto(_) => tail,
      _ => CTail::Goto(self.add_block(tail)),
    }
  }
}

pub fn explicate_control(mut prog: Program<IdxVar, Type>) -> CProgram<CInfo> {
  let locals = collect_locals(&prog);
  let mut state = State {
    blocks: vec![],
    label_index: 0,
  };
  let start = explicate_tail(&mut state, prog.body.pop().unwrap());
  state.blocks.insert(0, (Label::Start, start));

  let blocks = remove_unreachable_blocks(state.blocks);

  CProgram {
    info: CInfo { locals },
    body: blocks,
    types: prog.types,
  }
}

pub fn collect_locals(prog: &Program<IdxVar, Type>) -> IndexSet<IdxVar> {
  let mut locals = IndexSet::new();
  for exp in &prog.body {
    collect_exp_locals(exp, &mut locals);
  }
  locals
}

fn collect_exp_locals(exp: &Exp<IdxVar, Type>, locals: &mut IndexSet<IdxVar>) {
  match &exp.kind {
    ExpKind::Int(_)
    | ExpKind::Var(_)
    | ExpKind::Str(_)
    | ExpKind::Bool(_)
    | ExpKind::Void
    | ExpKind::Get(_) => {}
    ExpKind::Let { var, init, body } => {
      if init.ty != Type::Void {
        locals.insert(var.1.clone());
      }
      collect_exp_locals(&*init, locals);
      collect_exp_locals(&*body, locals);
    }
    ExpKind::Prim { op: _, args } => {
      for arg in args {
        collect_exp_locals(arg, locals);
      }
    }
    ExpKind::If { cond, conseq, alt } => {
      collect_exp_locals(&*cond, locals);
      collect_exp_locals(&*conseq, locals);
      collect_exp_locals(&*alt, locals);
    }
    ExpKind::Set { var: _, exp } => {
      collect_exp_locals(&*exp, locals);
    }
    ExpKind::Begin { seq, last } => {
      for exp in seq {
        collect_exp_locals(&*exp, locals);
      }
      collect_exp_locals(&*last, locals);
    }
    ExpKind::While { cond, body } => {
      collect_exp_locals(&*cond, locals);
      collect_exp_locals(&*body, locals);
    }
    ExpKind::Print(args) => {
      for arg in args {
        collect_exp_locals(arg, locals);
      }
    }
    ExpKind::NewLine => {}
  }
}

fn remove_unreachable_blocks(
  blocks: Vec<(Label, CTail)>,
) -> Vec<(Label, CTail)> {
  let mut blocks: IndexMap<_, _> = blocks.into_iter().collect();
  let mut reachable_blocks = IndexMap::<Label, CTail>::default();
  let mut worklist = VecDeque::new();
  worklist.push_back((Label::Start, blocks.remove(&Label::Start).unwrap()));

  while let Some((label, block)) = worklist.pop_front() {
    let mut tail = &block;
    loop {
      match tail {
        CTail::Seq(_, t) => {
          tail = &**t;
          continue;
        }
        CTail::Goto(goto_label) => {
          if let Some(goto_block) = blocks.remove(goto_label) {
            worklist.push_back((*goto_label, goto_block));
          }
        }
        CTail::If {
          conseq: goto_label1,
          alt: goto_label2,
          ..
        } => {
          if let Some(goto_block) = blocks.remove(goto_label1) {
            worklist.push_back((*goto_label1, goto_block));
          }
          if let Some(goto_block) = blocks.remove(goto_label2) {
            worklist.push_back((*goto_label2, goto_block));
          }
        }
        _ => {}
      }
      break;
    }
    reachable_blocks.insert(label, block);
  }

  reachable_blocks.into_iter().collect()
}

fn explicate_tail(state: &mut State, exp: Exp<IdxVar, Type>) -> CTail {
  match exp.kind {
    ExpKind::Int(_)
    | ExpKind::Var(_)
    | ExpKind::Bool(_)
    | ExpKind::Str(_)
    | ExpKind::Void
    | ExpKind::Get(_) => CTail::Return(CExp::Atom(atom(exp))),
    ExpKind::Prim {
      op: (_, "vector-set!"),
      ..
    } => {
      explicate_exp_effect(state, exp, CTail::Return(CExp::Atom(CAtom::Void)))
    }
    // TODO this should be removed in chapter 9
    ExpKind::Prim {
      op: (_, "vector-length"),
      mut args,
    } => match args.pop().unwrap().ty {
      Type::Vector(fields) => {
        CTail::Return(CExp::Atom(CAtom::Int(fields.len() as i64)))
      }
      Type::Alias(_) => todo!(),
      _ => unreachable!(),
    },
    ExpKind::Prim { op: (_, op), args } => {
      CTail::Return(CExp::Prim(prim(op, args)))
    }
    ExpKind::Let { var, init, body } => {
      let cont = explicate_tail(state, *body);
      explicate_assign(state, var.1, *init, cont)
    }
    ExpKind::If { cond, conseq, alt } => {
      let conseq = explicate_tail(state, *conseq);
      let alt = explicate_tail(state, *alt);
      explicate_pred(state, *cond, conseq, alt)
    }
    ExpKind::Begin { seq, last } => {
      let cont = explicate_tail(state, *last);
      explicate_effect(state, seq.into_iter(), cont)
    }
    ExpKind::While { .. }
    | ExpKind::Set { .. }
    | ExpKind::Print { .. }
    | ExpKind::NewLine => {
      explicate_exp_effect(state, exp, CTail::Return(CExp::Atom(CAtom::Void)))
    }
  }
}

fn explicate_assign(
  state: &mut State,
  var: IdxVar,
  init: Exp<IdxVar, Type>,
  cont: CTail,
) -> CTail {
  match init.kind {
    ExpKind::Int(_) | ExpKind::Bool(_) | ExpKind::Str(_) => {
      let assign = CStmt::Assign {
        var,
        exp: CExp::Atom(atom(init)),
      };
      CTail::Seq(assign, box cont)
    }
    ExpKind::Var(_) | ExpKind::Get(_) => {
      if init.ty == Type::Void {
        cont
      } else {
        let assign = CStmt::Assign {
          var,
          exp: CExp::Atom(atom(init)),
        };
        CTail::Seq(assign, box cont)
      }
    }
    ExpKind::Void => cont,
    ExpKind::Prim {
      op: (_, "vector-set!"),
      ..
    } => explicate_exp_effect(state, init, cont),
    ExpKind::Prim {
      op: (_, "vector-length"),
      mut args,
    } => match args.pop().unwrap().ty {
      Type::Vector(fields) => {
        let assign = CStmt::Assign {
          var,
          exp: CExp::Atom(CAtom::Int(fields.len() as i64)),
        };
        CTail::Seq(assign, box cont)
      }
      Type::Alias(_) => todo!(),
      _ => unreachable!(),
    },
    ExpKind::Prim { op: (_, op), args } => {
      if init.ty == Type::Void {
        cont
      } else {
        let prim = prim(op, args);
        let assign = CStmt::Assign {
          var,
          exp: CExp::Prim(prim),
        };
        CTail::Seq(assign, box cont)
      }
    }
    ExpKind::Let {
      var: (_, var1),
      init: init1,
      body,
    } => {
      let cont = explicate_assign(state, var, *body, cont);
      explicate_assign(state, var1, *init1, cont)
    }
    ExpKind::If { cond, conseq, alt } => {
      let cont = state.tail_to_goto(cont);
      let conseq = explicate_assign(state, var.clone(), *conseq, cont.clone());
      let alt = explicate_assign(state, var, *alt, cont);
      explicate_pred(state, *cond, conseq, alt)
    }
    ExpKind::Begin { seq, last } => {
      let cont = explicate_assign(state, var, *last, cont);
      explicate_effect(state, seq.into_iter(), cont)
    }
    ExpKind::While { .. }
    | ExpKind::Print { .. }
    | ExpKind::NewLine
    | ExpKind::Set { .. } => explicate_exp_effect(state, init, cont),
  }
}

fn explicate_pred(
  state: &mut State,
  cond: Exp<IdxVar, Type>,
  conseq: CTail,
  alt: CTail,
) -> CTail {
  match cond.kind {
    ExpKind::Bool(v) => {
      if v {
        conseq
      } else {
        alt
      }
    }
    ExpKind::Var(_) | ExpKind::Get(_) => {
      let conseq = state.tail_to_label(conseq);
      let alt = state.tail_to_label(alt);
      CTail::If {
        cmp: CCmpOp::Eq,
        lhs: atom(cond),
        rhs: CAtom::Bool(false),
        conseq: alt,
        alt: conseq,
      }
    }
    ExpKind::Let { var, init, body } => {
      let cont = explicate_pred(state, *body, conseq, alt);
      explicate_assign(state, var.1, *init, cont)
    }
    ExpKind::Prim {
      op: (_, "not"),
      mut args,
    } => explicate_pred(state, args.pop().unwrap(), alt, conseq),
    ExpKind::Prim { op, mut args } => {
      let conseq = state.tail_to_label(conseq);
      let alt = state.tail_to_label(alt);
      let cmp = op.1.parse().unwrap();
      let rhs = args.pop().unwrap();
      let lhs = args.pop().unwrap();
      CTail::If {
        cmp,
        lhs: atom(lhs),
        rhs: atom(rhs),
        conseq,
        alt,
      }
    }
    ExpKind::If {
      cond: cond1,
      conseq: conseq1,
      alt: alt1,
    } => {
      let conseq = state.tail_to_goto(conseq);
      let alt = state.tail_to_goto(alt);
      let conseq1 =
        explicate_pred(state, *conseq1, conseq.clone(), alt.clone());
      let alt1 = explicate_pred(state, *alt1, conseq, alt);
      explicate_pred(state, *cond1, conseq1, alt1)
    }
    ExpKind::Begin { seq, last } => {
      let cont = explicate_pred(state, *last, conseq, alt);
      explicate_effect(state, seq.into_iter(), cont)
    }
    ExpKind::Void
    | ExpKind::Int(_)
    | ExpKind::Str(_)
    | ExpKind::While { .. }
    | ExpKind::Print { .. }
    | ExpKind::NewLine
    | ExpKind::Set { .. } => {
      unreachable!()
    }
  }
}

fn explicate_effect(
  state: &mut State,
  seq: impl DoubleEndedIterator<Item = Exp<IdxVar, Type>>,
  cont: CTail,
) -> CTail {
  seq.rfold(cont, |cont, exp| explicate_exp_effect(state, exp, cont))
}

fn explicate_exp_effect(
  state: &mut State,
  exp: Exp<IdxVar, Type>,
  cont: CTail,
) -> CTail {
  match exp.kind {
    ExpKind::Bool(_)
    | ExpKind::Int(_)
    | ExpKind::Var(_)
    | ExpKind::Str(_)
    | ExpKind::Void
    | ExpKind::Get(_) => cont,
    ExpKind::Prim {
      op: (_, "read"),
      args: _,
    } => CTail::Seq(CStmt::Read, box cont),
    ExpKind::Prim {
      op: (_, "vector-set!"),
      mut args,
    } => {
      let val = args.pop().unwrap();
      let index = args.pop().unwrap();
      let vec = args.pop().unwrap();
      match index.kind {
        ExpKind::Int(index) => match &vec.ty {
          Type::Vector(fields) => {
            let fields_before = fields[..index as usize].to_vec();
            CTail::Seq(
              CStmt::VecSet {
                vec: atom(vec),
                fields_before,
                val: atom(val),
              },
              box cont,
            )
          }
          Type::Alias(_) => todo!(),
          _ => unreachable!(),
        },
        _ => unreachable!(),
      }
    }
    ExpKind::Prim { op: _, args } => {
      explicate_effect(state, args.into_iter(), cont)
    }
    ExpKind::Let { var, init, body } => {
      let body = explicate_exp_effect(state, *body, cont);
      explicate_assign(state, var.1, *init, body)
    }
    ExpKind::If { cond, conseq, alt } => {
      let cont = state.tail_to_goto(cont);
      let conseq = explicate_exp_effect(state, *conseq, cont.clone());
      let alt = explicate_exp_effect(state, *alt, cont);
      explicate_pred(state, *cond, conseq, alt)
    }
    ExpKind::Set { var, exp } => explicate_assign(state, var.1, *exp, cont),
    ExpKind::Begin { seq, last } => {
      let cont = explicate_exp_effect(state, *last, cont);
      explicate_effect(state, seq.into_iter(), cont)
    }
    ExpKind::While { cond, body } => {
      let loop_start = state.new_label();
      let body = explicate_exp_effect(state, *body, CTail::Goto(loop_start));
      let block = explicate_pred(state, *cond, body, cont);
      state.add_label_block(loop_start, block);
      CTail::Goto(loop_start)
    }
    ExpKind::NewLine => CTail::Seq(CStmt::NewLine, box cont),
    ExpKind::Print(args) => args.into_iter().rfold(
      CTail::Seq(CStmt::NewLine, box cont),
      |cont, exp| {
        CTail::Seq(
          CStmt::Print {
            ty: exp.ty.clone(),
            val: atom(exp),
          },
          box cont,
        )
      },
    ),
  }
}

fn atom(exp: Exp<IdxVar, Type>) -> CAtom {
  match exp.kind {
    ExpKind::Int(n) => CAtom::Int(n),
    ExpKind::Var(var) | ExpKind::Get(var) => CAtom::Var(var),
    ExpKind::Bool(v) => CAtom::Bool(v),
    ExpKind::Str(s) => CAtom::Str(s),
    ExpKind::Void => CAtom::Void,
    _ => unreachable!("{:?}", exp),
  }
}

fn prim(op: &str, mut args: Vec<Exp<IdxVar, Type>>) -> CPrim {
  match op {
    "read" => CPrim::Read,
    "-" => {
      if args.len() == 1 {
        let arg = args.pop().unwrap();
        CPrim::Neg(atom(arg))
      } else {
        let arg2 = args.pop().unwrap();
        let arg1 = args.pop().unwrap();
        CPrim::Sub(atom(arg1), atom(arg2))
      }
    }
    "+" => {
      let arg2 = args.pop().unwrap();
      let arg1 = args.pop().unwrap();
      CPrim::Add(atom(arg1), atom(arg2))
    }
    "not" => {
      let arg = args.pop().unwrap();
      CPrim::Not(atom(arg))
    }
    "eq?" | ">" | ">=" | "<" | "<=" => {
      let arg2 = args.pop().unwrap();
      let arg1 = args.pop().unwrap();
      CPrim::Cmp(op.parse().unwrap(), atom(arg1), atom(arg2))
    }
    "vector" => CPrim::Vector(
      args
        .into_iter()
        .map(|arg| {
          let ty = arg.ty.clone();
          (atom(arg), ty)
        })
        .collect(),
    ),
    "vector-ref" => {
      let index = args.pop().unwrap();
      let vec = args.pop().unwrap();
      match index.kind {
        ExpKind::Int(index) => {
          let ty = vec.ty.clone();
          match ty {
            Type::Vector(fields) => CPrim::VecRef {
              vec: atom(vec),
              fields_before: fields[..index as usize].to_vec(),
            },
            Type::Alias(_) => todo!(),
            _ => unreachable!(),
          }
        }
        _ => unreachable!(),
      }
    }
    "vector-length" => {
      let arg = args.pop().unwrap();
      CPrim::VecLen(atom(arg))
    }
    _ => unreachable!("{}", op),
  }
}

#[cfg(test)]
mod tests {
  use super::super::*;
  use super::*;
  use ch2::pass::remove_complex_operands;
  use insta::assert_snapshot;

  #[test]
  fn let_in_init() {
    let prog =
      ast::parse(r#"(let ([y (let ([x 20]) (+ x (let ([x 22]) x)))]) y)"#)
        .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn nested_prims() {
    let prog = ast::parse(
      r#"(let ([x (read)] [y (+ 2 3)]) (+ (- (read)) (+ y (- 2))))"#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn if_form() {
    let prog =
      ast::parse(r#"(if (> 3 (read)) (- 7 2) (- (let ([x 3]) (+ x (read)))))"#)
        .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn nested_if() {
    let prog = ast::parse(
      r#"
(let ([x (read)]
      [y (read)])
  (if (if (< x 1)
        (eq? x 0)
        (eq? x 2))
    (+ y 2)
    (+ y 10)))
"#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn const_cond() {
    let prog = ast::parse(r#"(if (not #t) 1 2)"#).unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn if_in_init() {
    let prog = ast::parse(
      r#"(let ([x (if (>= (read) 3) 10 77)]) (if (not (eq? x 10)) 41 2))"#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn var_cond() {
    let prog =
      ast::parse(r#"(let ([x (eq? (read) 20)]) (if x 42 89))"#).unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn shrink_and() {
    let prog =
      ast::parse(r#"(if (and (eq? (read) 0) (eq? (read) 2)) 0 42)"#).unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
