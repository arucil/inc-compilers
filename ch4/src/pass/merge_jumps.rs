use asm::{Program, Label, Instr, Block};
use indexmap::IndexMap;

pub fn merge_jumps<INFO>(prog: Program<INFO>) -> Program<INFO> {
  let mut refs = IndexMap::<Label, usize>::new();
  let mut blocks = IndexMap::<Label, Block>::new();
  for (label, block) in prog.blocks {
    if let Instr::Jmp(label) = block.code.last().unwrap() {
      *refs.entry(label.clone()).or_default() += 1;
      if block.code.len() > 1 {
        if let Instr::Jmp(label) = &block.code[block.code.len() - 2] {
          *refs.entry(label.clone()).or_default() += 1;
        }
      }
    }
    blocks.insert(label, block);
  }
  let mut label = Label::EntryPoint;
  let mut block = blocks.remove(&label).unwrap();
  let mut new_blocks = vec![];
  while let Instr::Jmp(next_label) = block.code.last().unwrap() {
    let next_label = *next_label;
    if refs[&next_label] == 1 {
      let mut next_block = blocks.remove(&next_label).unwrap();
      block.code.pop();
      block.code.append(&mut next_block.code);
    } else {
      block.code.pop();
      new_blocks.push((label, block));
      label = next_label;
      block = blocks.remove(&label).unwrap();
    }
  }
  new_blocks.push((label, block));
  new_blocks.extend(blocks.into_iter());
  Program {
    info: prog.info,
    blocks: new_blocks,
  }
}