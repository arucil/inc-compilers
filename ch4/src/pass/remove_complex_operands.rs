#[cfg(test)]
mod tests {
  use super::super::*;
  use ch2::pass::remove_complex_operands;
  use ast::*;
  use insta::assert_snapshot;

  #[test]
  fn nested_prims() {
    let prog =
      parse(r#"(let ([x (read)] [y (+ 2 3)]) (+ (- (read)) (+ y (- 2))))"#)
        .unwrap();
    let prog = uniquify::uniquify(prog);
    let result = remove_complex_operands::remove_complex_operands(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn init_with_var() {
    let prog = parse(r#"(let ([a 42]) (let ([b a]) b))"#).unwrap();
    let prog = uniquify::uniquify(prog);
    let result = remove_complex_operands::remove_complex_operands(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn let_in_init() {
    let prog = parse(
      r#"(let ([a (+ (let ([x (read)] [y (- x)]) (+ x (- y))) 7)]) (- a))"#,
    )
    .unwrap();
    let prog = uniquify::uniquify(prog);
    let result = remove_complex_operands::remove_complex_operands(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn if_form() {
    let prog = parse(
      r#"(+ (- 3 2) (if (not (eq? 2 3)) (- (read)) (> (read) (not #t))))"#,
    )
    .unwrap();
    let prog = uniquify::uniquify(prog);
    let result = remove_complex_operands::remove_complex_operands(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn begin() {
    let prog = parse(
      r#"
(let ([x0 10])
  (let ([y1 0])
    (+ (+ (begin (set! y1 (read)) x0)
       (begin (set! x0 (read)) y1))
    x0)))"#,
    )
    .unwrap();
    let prog = uniquify::uniquify(prog);
    let result = remove_complex_operands::remove_complex_operands(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
