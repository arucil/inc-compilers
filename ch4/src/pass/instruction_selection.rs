use super::explicate_control::CInfo;
use asm::Program;
use ast::{IdxVar, Type};
use ch2::pass::instruction_selection::{CodeGen, Locals};
use control::*;
use indexmap::IndexMap;
use std::fmt::{self, Display, Formatter};

#[derive(Default)]
pub struct Info {
  pub locals: IndexMap<IdxVar, Type>,
}

impl Display for Info {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    writeln!(f, "locals: {:?}\n", self.locals)
  }
}

impl From<CInfo> for Info {
  fn from(info: CInfo) -> Self {
    Self {
      locals: info.locals,
    }
  }
}

impl Locals for Info {
  fn add_local(&mut self, local: IdxVar, ty: Type) {
    self.locals.insert(local, ty);
  }
}

pub fn select_instruction(
  prog: CProgram<CInfo>,
  use_heap: bool,
) -> Program<Info, IdxVar> {
  let types = prog.types;
  let mut codegen = CodeGen::new(&types, use_heap);
  let funs = prog
    .funs
    .into_iter()
    .map(|fun| codegen.gen_fun(fun))
    .collect();
  let blocks = codegen.gen_body(prog.body);
  let result = codegen.finish();
  Program {
    info: Info {
      locals: prog.info.locals,
    },
    blocks,
    constants: result.constants,
    externs: result.externs,
    funs,
    types,
  }
}

#[cfg(test)]
mod tests {
  use super::super::*;
  use super::*;
  use ast::*;
  use ch2::pass::remove_complex_operands;
  use insta::assert_snapshot;

  #[test]
  fn nested_prims() {
    let prog =
      parse(r#"(let ([x (read)] [y (+ 2 3)]) (+ (- (read)) (+ y (- 2))))"#)
        .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let prog = explicate_control::explicate_control(prog);
    let result = select_instruction(prog, false);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn nested_if() {
    let prog = ast::parse(
      r#"
(let ([x (read)]
      [y (read)])
  (if (if (< x 1)
        (eq? x 0)
        (eq? x 2))
    (+ y 2)
    (+ y 10)))
    "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let prog = explicate_control::explicate_control(prog);
    let result = select_instruction(prog, false);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn if_in_init() {
    let prog = ast::parse(
      r#"(let ([x (if (>= (read) 3) 10 77)]) (if (not (eq? x 10)) 41 2))"#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let prog = explicate_control::explicate_control(prog);
    let result = select_instruction(prog, false);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn complex_if() {
    let prog = ast::parse(
      r#"
(let ([x (read)])
  (if (> x 100)
    (let ([y (read)])
      (if (or (and (>= x y) (< (- x y) 10))
              (and (< x y) (< (- y x) 10)))
        -1
        (- y)))
    (if (and (> x 40) (< x 60))
      5000
      x)))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let prog = explicate_control::explicate_control(prog);
    let result = select_instruction(prog, false);
    assert_snapshot!(result.to_string_pretty());
  }
}
