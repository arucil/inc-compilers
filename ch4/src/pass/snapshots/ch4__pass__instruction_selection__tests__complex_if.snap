---
source: ch4/src/pass/instruction_selection.rs
expression: result.to_string_pretty()
---
locals: {x.0: Int, y.1: Int, (tmp).0: Int, (tmp).1: Int}

constants: {}

.start:
    call rt_read_int
    mov x.0, rax
    cmp x.0, 100
    jg .block8
    jmp .block9
.block8:
    call rt_read_int
    mov y.1, rax
    cmp x.0, y.1
    jge .block4
    jmp .block3
.block9:
    cmp x.0, 40
    jg .block7
    jmp .block6
.block4:
    mov (tmp).0, x.0
    sub (tmp).0, y.1
    cmp (tmp).0, 10
    jl .block0
    jmp .block3
.block3:
    cmp x.0, y.1
    jl .block2
    jmp .block1
.block7:
    cmp x.0, 60
    jl .block5
    jmp .block6
.block6:
    mov rax, x.0
    jmp .epilogue
.block0:
    mov rax, -1
    jmp .epilogue
.block2:
    mov (tmp).1, y.1
    sub (tmp).1, x.0
    cmp (tmp).1, 10
    jl .block0
    jmp .block1
.block1:
    mov rax, y.1
    neg rax
    jmp .epilogue
.block5:
    mov rax, 5000
    jmp .epilogue

